rust/src/coordinator_async/message_processing.rs

```diff
@@ -0,0 +1,278 @@
+use crate::{
+    coordinator::RoundSeed,
+    crypto::ByteObject,
+    message::{Sum2Owned, SumOwned, UpdateOwned},
+    ParticipantPublicKey,
+    ParticipantTaskSignature,
+    PetError,
+    SumParticipantPublicKey,
+    UpdateParticipantPublicKey,
+};
+use std::{future::Future, sync::Arc};
+use tokio::{
+    sync::{
+        broadcast,
+        mpsc::{unbounded_channel, Sender, UnboundedReceiver, UnboundedSender},
+    },
+    time::Duration,
+};
+
+pub struct SumValidationData {
+    pub sum: f64,
+    pub seed: RoundSeed,
+}
+
+pub struct UpdateValidationData {
+    pub sum: f64,
+    pub update: f64,
+    pub seed: RoundSeed,
+}
+
+// A sink to collect the results of the MessageHandler tasks.
+pub struct MessageSink {
+    // The minimum number of successfully validated messages.
+    min_messages: usize,
+    // A counter that is incremented when a message has been successfully validated.
+    successful_messages: usize,
```

Comment:
i'm still undecided if we should actually short-circuit a phase. selection bias can have a negative impact on the model quality and this choice of aborting earlier than the `max_duration` can have various consequences:
- update participants with less training data will probably be faster and therefore have a higher representation in the aggregated model, even though they contribute less
- malicious update participants who just submit garbage immediately will have a higher representation in the aggregated model
- update participants from different parts of the world with slower internet will have a lower representation in the aggregated model, even though the model would benefit from more different and less similar data points
- etc

basically, the `sum`/`update` thresholds of the protocol give a `mean_messages` value we can expect, therefore we should only check that the `successful_messages` are greater than the minimum number of message required for a save protocol and otherwise just wait for the timeout. this was perviously called `min_sum`/`min_update` in the `Coordinator`, but i'm not sure if the `min_messages` of the `MessageSink` is supposed to have the same meaning?